internal/engine/portforwardcontroller.go

```diff
@@ -100,26 +102,73 @@ func (m *PortForwardController) OnChange(ctx context.Context, st store.RStore) {
 	}
 
 	for _, entry := range toStart {
-		entry := entry
-		ns := entry.namespace
-		podID := entry.podID
+		// Treat port-forwarding errors as part of the pod log
+		actionWriter := PodLogActionWriter{
+			store:        st,
+			podID:        entry.podID,
+			manifestName: entry.name,
+		}
+		ctx := logger.WithLogger(ctx, logger.NewLogger(logger.Get(ctx).Level(), actionWriter))
+
 		for _, forward := range entry.forwards {
-			// TODO(nick): Handle the case where DockerForDesktop is handling
-			// the port-forwarding natively already
-			_, closer, err := m.kClient.ForwardPort(ctx, ns, podID, forward.LocalPort, forward.ContainerPort)
-			if err != nil {
-				logger.Get(ctx).Infof("Error port-forwarding %s: %v", entry.name, err)
-				continue
-			}
+			entry := entry
+			forward := forward
+			go m.startPortForwardLoop(ctx, entry, forward)
+		}
+	}
+}
+
+func (m *PortForwardController) startPortForwardLoop(ctx context.Context, entry portForwardEntry, forward model.PortForward) {
+	originalBackoff := wait.Backoff{
+		Steps:    1000,
+		Duration: 50 * time.Millisecond,
+		Factor:   2.0,
+		Jitter:   0.1,
+		Cap:      15 * time.Second,
+	}
+	currentBackoff := originalBackoff
+
+	for {
+		start := time.Now()
+		err := m.onePortForward(ctx, entry, forward)
+		if ctx.Err() != nil {
+			// If the context was canceled, we're satisfied.
+			// Ignore any errors.
+			return
+		}
+
+		// Otherwise, repeat the loop, maybe logging the error
+		if err != nil {
+			logger.Get(ctx).Infof("Error port-forwarding %s: %v", entry.name, err)
+		}
 
-			go func() {
-				<-entry.ctx.Done()
-				closer()
-			}()
+		// If this failed in less than a second, then we should advance the backoff.
+		// Otherwise, reset the backoff.
+		if time.Since(start) < time.Second {
+			time.Sleep(currentBackoff.Step())
+		} else {
+			currentBackoff = originalBackoff
 		}
 	}
 }
 
+func (m *PortForwardController) onePortForward(ctx context.Context, entry portForwardEntry, forward model.PortForward) error {
+	ns := entry.namespace
+	podID := entry.podID
+
+	// TODO(nick): Add exponential retry & backoff
```

Comment:
is this still relevant, or is it taken care of by startPortForwardLoop?

Reply:
done